DietPi-CloudShell | Run nyx on a separate TTYl

nonnorm opened the pull request:
**Status**: Ready

**Commit list/description**:
- DietPi-CloudShell | Instead of killing and restarting nyx, keep the process alive on a different TTYI and switch back and forth

nonnorm committed:
Change nyx implementation

MichaIng commented on dietpi/dietpi-cloudshell:
```diff
@@ -33,6 +33,9 @@
 	readonly FP_TMP='.tmp'
+	
+	# Allow Job Control
+	set -m
```
Let's set this only when the scene is enabled. E.g. set it below in the `if [[ ${aEnabledScenes[9]} == 1 ]]; then` block.

MichaIng commented on dietpi/dietpi-cloudshell:
```diff
@@ -1054,11 +1057,11 @@
+	# Tor Relay: Start displaying if not already
 	NYX_ACTIVE=
-	Update_Scene_9(){ [[ $NYX_ACTIVE ]] || { clear; nyx & NYX_ACTIVE=1; }; }
+	Update_Scene_9(){ [[ $NYX_ACTIVE ]] || { clear; fg & NYX_ACTIVE=1; }; }
```
Just in case there is any other job sent to background, the wrong one would be pulled to foreground 🤔.

MichaIng commented on dietpi/dietpi-cloudshell:
```diff
@@ -1196,6 +1199,12 @@ _EOF_
 
+		# Start Tor Relay monitoring if the scene is active
+		if [[ ${aEnabledScenes[9]} == 1 ]]; then
+			nyx &
```
Let's store the PID in a variable:
```
NYX_PID=$!
```
Then we can use `fg $NYX_PID` above and `kill -STOP $NYX_PID` below and in exit trap
```
[[ $NYX_PID ]] && kill $NYX_PID
```
in the exit trap.

This has the little bonus that in theory multiple `nyx` (or `dietpi-cloudshell`) processes can run and we are sure to only kill and control the one that was called within this instance.

nonnorm committed:
Fixes

nonnorm committed:
CodeFactor fixes

nonnorm committed:
Add missing ;

nonnorm committed:
Fix quotes

nonnorm commented:
Strange, even after running `set -m` job control is disabled.

MichaIng commented:
Indeed, even if `set -m` is placed directly before `fg` 🤔. I'll merge v6.34.1 now, but we'll do another loop and find out the reason and fix for this.

nonnorm commented:
It does change `$SHELLOPTS` like it is supposed to, adding `:monitor` to the end.  It seems more like a problem with `fg` not reading it correctly.

MichaIng commented:
```
hmB braceexpand:hashall:interactive-comments:monitor
/boot/dietpi/dietpi-cloudshell: line 1065: fg: no job control
```
`$-` as well has the `m` flag.

MichaIng commented:
`fg &` is the problem, this combination seems to be not allowed, although the error message is strange. Removing the immediate sent to background makes it succeeding, but of course the script halts then.

Another issue: `fg <PID>` is not correct, but `fg %<Job_ID>` is, means the IDs that you in the square brackets when handling background jobs:
```
2020-12-12 13:49:04 root@VM-Buster:/tmp# sleep 1000 &
[4] 2865
[3]   Exit 1                  fg
2020-12-12 13:50:57 root@VM-Buster:/tmp# jobs
[1]   Running                 sleep 1000 &
[2]-  Running                 sleep 1000 &
[4]+  Running                 sleep 1000 &
```
The problem is that I could not find any variable to get the job ID of the last created background job... Also I'm not sure if it stays the same when pulling it to foreground and sending to background again, only via ugly multiple commands and parsing back and forth:
```
2020-12-12 13:51:01 root@VM-Buster:/tmp# jobs -l
[1]   2858 Running                 sleep 1000 &
[2]-  2863 Running                 sleep 1000 &
[4]+  2865 Running                 sleep 1000 &
```

So I think, job control is not really what we need here. We do not want to send a job from background to foreground and back (which seems to be not possible), but the job must stay a background job (what we tried via `fg &`) so that the script can continue, but only it must stop and continue to send its output to the console, doesn't it? Probably we can simply change the STDOUT (+STDERR) from current console to /dev/null and back. But even that seems to be not easily possible:
```
l-wx------ 1 root root 64 Dec 12 14:09 /proc/13975/fd/1 -> /dev/null
2020-12-12 14:15:33 root@VM-Buster:~# ln -sf /dev/stdout /proc/13975/fd/1
ln: failed to create symbolic link '/proc/13975/fd/1': No such file or directory
2020-12-12 14:15:41 root@VM-Buster:~# ln -sf /dev/pts/0 /proc/13975/fd/1
ln: failed to create symbolic link '/proc/13975/fd/1': No such file or directory
```

But hey, since we use SIGSTOP, what about SIGCONT to continue the process? https://en.wikipedia.org/wiki/Job_control_(Unix)#Implementation
Testing...

nonnorm commented:
From when I tried SIGCONT, it just resumed `nyx` without printing anything to the screen.  Same with `bg`.  I didn't actually try it in the script, though.

MichaIng commented:
It works here, but somehow the output is malformed, parts are missing, the following scenes output is scattered across the screen 🤔.
**EDIT: It manipulates the terminal, `tput reset` or `stty sane` can be used to bring everything back to normal for other scenes. However you are right that continuing nyx only shows updated characters but does not print all the static info/format.**

MichaIng commented:
Hmm, actually I don't see a way to keep `nyx` running while turning on and off output as we need it. It configures the terminal on startup and prints static content only once and when the screen is cleared afterwards, the static content is missing. I tried `tput init`/`tput reset` to trigger a redraw, but while one can see the screen turning black shortly, nyx does not redraw static content.

Looks like killing and restarting nyx is the best we can do, but I agree it is not great since the graph is always reset and it takes ~1 second for nyx startup 😢.

nonnorm commented:
Yeah, it also wreaks havoc on the logs.

MichaIng committed:
Merge branch 'dev' into nyx-fix

MichaIng committed:
v6.34

+ DietPi-CloudShell | Tor monitor: Fixes due to conflict resolving and stop nyx before checking and probably entering blank mode

MichaIng committed:
v6.34

+ DietPi-CloudShell | Assure that nyx stopped as well before entering blank time

MichaIng committed:
Merge branch 'dev' into nyx-fix

MichaIng commented:
Okay, I updated dev branch to include the nyx PID-based handling and another fix where nyx would have continued even that the screen entered a specified blank time, so that this PR only includes the essential changes to keep it running, even that it does not yet work. Probably we find a way later.

Btw, ignore the shellcheck error. No idea why it shows up here while it does not show up for dev branch. Since NYX_PID only ever declared as empty string or assigned a PID, shellcheck should actually be smart enough to see that no issue can occur. Otherwise it can be quoted in dev branch later.
_____________
Ah I just recognised that the traffic graph does not start from zero when nyx is started but already shows the whole traffic for the last minute. That's already much better than I though.

nonnorm committed:
Switch to the 'polite' signal

nonnorm commented:
It seems that stopping `nyx` with the 'polite' signal (`-TSTP`) stops it from being malformed (or malforming other things).  Also probably better anyway just to use that.

MichaIng commented:
Nice, that made it, I mean the whole nyx output is present again on CONT signal 👍. Format is still malformed. `tput reset` and `tput init` restore the original terminal size as well, so `stty sane` is better. It fails when there is no STDIN, but per option it can be bound to STDOUT.

MichaIng committed:
v6.34

+ DietPi-CloudShell | Create nyx killing exit trap only when+before nyx is actually started
+ DietPi-CloudShell | Reset sane output terminal settings, as nyx by times adjusts those so that following scenes have unexpected line indentations and/or overlap with nyx output remains.

MichaIng committed:
v6.34

+ DietPi-CloudShell | Avoid sending nyx stop command when it is not actually active

MichaIng commented:
The output is still malformed sometimes. Probably it is not good to send the kill -TSTP command directly affter command call. I'll merge it into the scene loop so that it can fully startup in foreground on its first turn. Probably `stty sane` is not required then, let's see.

nonnorm commented:
Just an (unrelated) thought, but what if we added `setfont /usr/share/consolefonts/DEFAULTFONT.psf` to the exit trap (or some other solution) to reset the font size if DietPi-CloudShell is quit?

MichaIng commented:
Yes, I thought the same, or store the previously used font and restore that. However, I'm now concentrating on nyx first.

MichaIng commented:
It's strange, the issues don't seem to have something to do with terminal reset, font size and all of this, but nyx simply sometimes startup in a broken state. Simply run `nyx` from console, sometimes it starts nice, all as it should be, sometimes you see smaller of larger of completely breaking artefacts. If this is the case, after closing nyx, also the console is in a broken shape, sometimes more, sometimes less.
From PuTTY:
![nyx](https://user-images.githubusercontent.com/28480705/102021345-78b1ef00-3d7f-11eb-9639-37668b0f1cf6.png)
From tty1:
![nyx](https://user-images.githubusercontent.com/28480705/102021597-15c15780-3d81-11eb-8425-7e3a5a8758f2.png)
......
Okay, just verified that it sometimes also enters this broken state while it worked well for a while.

Btw this is all already with start/stop within update loop.

nonnorm committed:
Move nyx starting to update loop

nonnorm committed:
Add wait after killing

nonnorm committed:
Remove waits

nonnorm commented:
What really seems to happen (to me at least) is that `nyx` doesn't stop when it's supposed to, leaving the updating graph and messing up everything else.

MichaIng commented:
On the "linux" terminal of VirtualBox, it fails on nearly every attempt.

> What really seems to happen (to me at least) is that nyx doesn't stop when it's supposed to, leaving the updating graph and messing up everything else.

But if it does not stop, it was already broken on startup or resume (SIGCONT), sometimes in a not notable way. As said, simply running nyx from the terminal sometimes shows flaws like in my first post above.

As long as it starts that unclean by times, we need to completely kill and restart it while looping. Not great but otherwise cloudshell will mess up earlier or later, not acceptable.

MichaIng commented:
While testing CloudShell up and down, I finally was brave enough to do a larger rework of some IMO strange flows it had: https://github.com/MichaIng/DietPi/pull/3974
However, it also revealed that nyx indeed does deeper terminal settings changes, e.g. while on startup, the script hides the cursor, but when killing nyx, the cursor reappears and also screen blanking/powersave is re-enabled. So we seem to require re-setting the terminal settings after killing nyx. I'm actually thinking, due to the nature of nyx, if it makes more sense to offer it as a dedicated autostart monitoring option, rather than baking it into DietPi-CloudShell. The latter consists of scenes with a single static print each loop, so standalone monitoring processes naturally are hard to implement, especially when we cannot control how they react to signals and what they to do the terminal settings.

MichaIng committed:
Merge branch 'dev' into nyx-fix

MichaIng committed:
v6.34

+ DietPi-CloudShell | Set nyx active when starting it the first time, else an obsolete SIGCONT would be sent

MichaIng commented:
I tried another approach:
- Redirecting nyx output to a file and paste the content of this file regularly.
- It works well when running the process in foreground on one smaller terminal and cat that file on another larger terminal.
- But running it within the same terminal as background process, the output is a complete mess. It seems to depend on STDOUT at the current terminal to detect it's attributes correctly, and configure it.

When using SIGTSTP, it halt sometimes, sometimes not, as it produces a clean output sometimes and sometimes not. Using SIGSTOP leads to static output being lost.

🈯 It does work perfectly fine when redirecting the nyx output to e.g. /dev/tty9 and using `chvt 9`/`chvt 1` to switch back and forth to that terminal. But that does only work when the originating terminal is already one of /dev/ttyN, i.e. not on SSH or similar pseudoterminal sessions and not on serial console (although no-one would attempt that I guess). I was looking for a way to create an additional pseudoterminal (e.g. /dev/pts/9) and connect an SSH session to it, but that does not seem to be possible.

What do you think, shall we disable nyx on all terminals which do not match `/dev/tty[1-9]` and print a warning about this on startup?

nonnorm commented:
Yes, I think that sounds like a good idea.  When you say 'startup', do you mean startup of DietPi-CloudShell?

MichaIng commented:
Startup of DietPi-CloudShell. TTY could be checked before entering the loop (where also the config file is pre-created), then print disable scene in case and print info. Since the tty is checked in other cases, it makes sense to do that once and store it to a variable.

MichaIng committed:
v6.34

+ DietPi-CloudShell | Tor Relay: Since nyx cannot be halted and resumed reliably, without breaking itself or the terminal, it prints now to a different TTY (/dev/tty9) and DietPi-CloudShell will switch to that TTY and away on scene switch. Since this does only work on a usual virtual terminal, not on pseudo terminal (/dev/pts/N) or serial terminals, it will be disabled there.

MichaIng committed:
v6.34

+ DietPi-CloudShell | Fix

MichaIng commented:
Finally, works now. @ravenclaw900 please give it a try.

A bid sad is that the nyx output still breaks after a while on my tests. But since it is on a different terminal, it does not effect the other sets anymore. Those issues need to be reported upstream then.

MichaIng committed:
v6.34

+ DietPi-CloudShell | Save nyx PID

MichaIng committed:
v6.34

+ DietPi-CloudShell | Optimise killing nyx on exit

nonnorm commented:
Good solution.  It works well.
